### ISSUE TITLE
MemoryError allocating large byte string values

### ISSUE THREAD

**Issue by hdkire** [numpy/numpy#19786]
Creating large byte string values can result in MemoryError and somewhat strange behavior.
It seems like they are limited to < 2 ** 31 characters, in my setup, is this the intended or expected behavior?

### Reproducing code example 1:

```python
import numpy as np
# OK
x = np.empty(1, dtype=f'S{2**30}')
x = np.empty(1, dtype=f'S{2**32}')
# Fail
x = np.empty(1, dtype=f'S{2**31}')
```

### Error message:

MemoryError: Unable to allocate -2.00 GiB for an array with shape (1,) and data type |S-2147483648

### Strange behavior:

When using even larger numbers, 2 ** 32 or more, there is no error but the resulting itemsize is 1.

###  Reproducing code example 2:

Reproduces also when tested on windows.

```python
import numpy as np
# OK
x = np.array(b'a' * (2 ** 30))
x = np.array(b'a' * (2 ** 31 - 1))

# Fail
x = np.array(b'a' * (2 ** 31))
```

### Error message:

MemoryError: Unable to allocate -2.00 GiB for an array with shape () and data type |S-2147483648

###  Reproducing code example 3:

```python
import numpy as np
# Fail
x = np.array(b'a' * (2 ** 32 ))
```

### Error message:

SystemError: <built-in function array> returned a result with an error set

### NumPy/Python version information:

1.20.3 3.9.5 (default, May 11 2021, 08:20:37) 
[GCC 10.3.0]

**Comment by behl1anmol** [comment]
Hey, I want to work on fixing the issue. I have started looking into the ndarray interfaces in numeric.py. Can you guide me further in looking into the issue?

**Comment by mattip** [comment]
What have you tried, what did you find? You probably are looking for a point in the C code where we use an int for a value that then overflows (int is 32-bits on windows).

**Comment by hdkire** [comment]
> What have you tried, what did you find? You probably are looking for a point in the C code where we use an int for a value that then overflows (int is 32-bits on windows).

I tested on both Windows and Linux (Ubuntu 21.04), and I guess that some 32-bit handling is used on Linux too. `2 ** 31 - 1` is the largest positive value for a 32-bit signed integer.

**Comment by eric-wieser** [comment]
This bug doesn't seem to happen for me in numpy 1.19.4 on windows.

**Comment by hdkire** [comment]
> This bug doesn't seem to happen for me in numpy 1.19.4 on windows.

The behavior on Windows was a bit different.

Code example 1: No exception, but the resulting itemsize remains at `2 ** 31 -1` for larger numbers. 

Code example 2  (`x = np.array(b'a' * (2 ** 31))`): MemoryError: Unable to allocate -2.00 GiB for an array with shape () and data type |S-2147483648.

Code example 3: No exception, but the resulting itemsize is 1.

### NumPy/Python version information (Windows)

1.19.4 3.8.6 (tags/v3.8.6:db45529, Sep 23 2020, 15:52:53) [MSC v.1927 64 bit (AMD64)]

**Comment by behl1anmol** [comment]
I guess that there is some 32-bit handling happening in 64-bit systems also within the codebase. I am using Ubuntu (20.04) and the Python shell is executing in 64-bit, still I get an error.

**Comment by behl1anmol** [comment]
> What have you tried, what did you find? You probably are looking for a point in the C code where we use an int for a value that then overflows (int is 32-bits on windows).

I am trying to find a part in the codebase wherein there is some 32-bit handling is explicitly due to which error is being raised on 64 bit systems also.

**Comment by seberg** [comment]
> It seems like they are limited to < 2 ** 31 characters, in my setup, is this the intended or expected behavior?

Yes, this limitation is correct.  Seems the descriptor code does not check for overflow to negative values on some platforms though, creating negative sizes that then end up raising the wrong error.

> I am trying to find a part in the codebase wherein there is some 32-bit handling is explicitly due to which error is being raised on 64 bit systems also.

The problem is the negative amount of characters here.  So you need to find where the string length is parsed and add a check (raise an error) there.

**Comment by mattip** [comment]
The reproducer now raises `TypeError: data type 'S4294967296' not understood`. I think this is "good enough" for this edge case. Closing.
